Add initial osg token renewer doc (SOFTWARE-4763) (#868)
opensciencegrid/docs: docs/other/osg-token-renewer.md

Installing and Using the OSG Token Renewal Service
==================================================

This document contains instructions to install and configure the
OSG Token Renewal Service package, `osg-token-renewer`,

> for obtaining and automatically renewing tokens with the OIDC tools.

Reviewer: ```suggestion
for obtaining and automatically renewing tokens with [oidc-agent](https://github.com/indigo-dc/oidc-agent).
```

Author: Ok, i thought to say OIDC tools because it uses the other `oidc-*` commands, but i guess the package is called `oidc-agent`, so not far off.

> This service automates the procedure in the
> [Requesting Tokens](https://opensciencegrid.org/technology/software/requesting-tokens/)
> document.

Reviewer: ```suggestion
```

Shh this doc is secret :)

`/osg/token-renewer/config.ini`, you need to create a corresponding `token`
section in the config for each token that should be generated for this account
(possibly with different options).

1.  Choose a `<TOKEN_NAME>` and add a new `[token <TOKEN_NAME>]` section
    (replacing the example section if it's still there).

    The `account` option in this section must match the `<ACCOUNT_SHORTNAME>`
    for the corresponding `[account <ACCOUNT_SHORTNAME>]` section.

1.  Set the `token_path` to `/etc/osg/tokens/<ACCOUNT_SHORTNAME>.token

1.  Optionally, you may also specify any of the following options, which will
    be passed to the `oidc-token` command when generating the token:

    | Option         | Description                                               |
    |:---------------|:----------------------------------------------------------|
    | `audience`     | list of audiences to pass via `--aud` option              |
    | `scope`        | list of scopes to pass via `--scope` option               |
    | `min_lifetime` | min token lifetime in seconds to pass via `--time` option |

    !!! note
        For tokens used against an HTCondor-CE, set the `audience` option to

>         `<CE FQDN>:<CE PORT>`.
> ### Example configuration

Reviewer: ```suggestion
        `<CE FQDN>:<CE PORT>`.

### Example configuration
```

   the client request with a passcode printed on the console.
   Follow the prompts (visit the web link, enter the request passcode,
   log in with your account for your issuer, and authorize the request).
1. If this succeeds, you will be prompted with a new
   `[account <ACCOUNT_SHORTNAME>]` section to add to your `config.ini`.
   Add the section to your `/etc/token-renewer/config.ini`,
   replacing the example section if it's still there.

Next you can configure one or more tokens for this client account.

### Configuring tokens

After you have created an OIDC client account and added it to
`/osg/token-renewer/config.ini`, you need to create a corresponding `token`
section in the config for each token that should be generated for this account
(possibly with different options).

1.  Choose a `<TOKEN_NAME>` and add a new `[token <TOKEN_NAME>]` section
    (replacing the example section if it's still there).

    The `account` option in this section must match the `<ACCOUNT_SHORTNAME>`
    for the corresponding `[account <ACCOUNT_SHORTNAME>]` section.

1.  Set the `token_path` to
    `/etc/osg/tokens/<ACCOUNT_SHORTNAME>.<TOKEN_NAME>.token`

> 1.  Optionally, you may also specify any of the following options, which will
>     be passed to the `oidc-token` command when generating the token:
> 
>     | Option         | Description                                               |
>     |:---------------|:----------------------------------------------------------|
>     | `audience`     | list of audiences to pass via `--aud` option              |
>     | `scope`        | list of scopes to pass via `--scope` option               |
>     | `min_lifetime` | min token lifetime in seconds to pass via `--time` option |

Reviewer: ```suggestion
1.  Optionally, you may also specify any of the following options, which can attenuate
    the respective values in the generated token compared to the associated with the account:

    | Option         | Description                                               |
    |:---------------|:----------------------------------------------------------|
    | `audience`     | list of audiences  (see [RFC7519](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3)) |
    | `scope`        | list of [scopes](https://github.com/WLCG-AuthZ-WG/common-jwt-profile/blob/master/profile.md#capability-based-authorization-scope)               |
    | `min_lifetime` | min token lifetime in seconds |
```

Author: > Optionally, you may also specify any of the following options, which can attenuate the respective values in the generated token compared to **the** associated **with the** account:

Should we change **the** to "those"?  Or drop the **with the** ?  It reads funny to me.

Also, "attenuate" is a big word to me that i had to look up.  Seems to mean "limit" ?

Reviewer: > Should we change the to "those"? Or drop the with the ? It reads funny to me.

Yeah, that's better

> Also, "attenuate" is a big word to me that i had to look up. Seems to mean "limit" ?

Sure, sounds good

    !!! note
        For tokens used against an HTCondor-CE, set the `audience` option to
        `<CE FQDN>:<CE PORT>`.

### Example configuration

> <insert example config here with account + token sections>

Reviewer: Do you have an example to post? Otherwise we should just drop this section

Author: Added.

Managing the OSG Token Renewal Service
--------------------------------------

These services are managed by `systemctl` and may start additional services as
dependencies.
As a reminder, here are common service commands (all run as `root`) for EL7:

| To...                                   | On EL7, run the command...         |
| :-------------------------------------- | :--------------------------------- |
| Start a service                         | `systemctl start <SERVICE-NAME>`   |
| Stop a service                          | `systemctl stop <SERVICE-NAME>`    |
| Enable a service to start on boot       | `systemctl enable <SERVICE-NAME>`  |
| Disable a service from starting on boot | `systemctl disable <SERVICE-NAME>` |

> ### Public cache services

Reviewer: ```suggestion
### Token renewal services
```

Installing and Using the OSG Token Renewal Service
==================================================

This document contains instructions to install and configure the
OSG Token Renewal Service package, `osg-token-renewer`,
for obtaining and renewing tokens with the OIDC tools.
This service automates the procedure in the
[Requesting Tokens](https://opensciencegrid.org/technology/software/requesting-tokens/)
document.

Before Starting
---------------

Before starting the installation process, consider the following points
(consulting [the Reference section below](#reference) as needed):

- **WLCG INDIGO IAM account:** You'll need an account with WLCG.
  You can register for an account

>   [here](https://wlcg.cloud.cnaf.infn.it/start-registration).

Author: In that case the script that shoves in the wlcg issuer endpoint for us when running oidc-gen is probably not general enough either...  I think i was going off the requesting tokens doc, which specifies the wlcg one in particular.

>   [here](https://wlcg.cloud.cnaf.infn.it/start-registration).

Reviewer: Apologies, that doc is mainly for internal usage, and I can count on getting everyone accounts at the WLCG IAM instance. We'll need to let users of the package to choose an issuer.

>   [here](https://wlcg.cloud.cnaf.infn.it/start-registration).

Reviewer: Nah, I think what you have is fine

-----------------------------------------

The main configuration file for the service is `/osg/token-renewer/config.ini`.

For each OIDC Client, you will add an `account` section to the config file.
For each token you wish to generate for this client account,
you will configure a `token` section with any relevant options.

Examples of this can be found in the `/osg/token-renewer/config.ini` that gets
installed with the package.

Each `[account <ACCOUNT_SHORTNAME>]` section corresponds to a client account
named `<ACCOUNT_SHORTNAME>`, set up with the `oidc-gen` tool, run by the
`osg-token-renewer-setup.sh` script.

In this `account` section, the `password_file` option is a path to a file
you create as `root` with the encryption password to be used for this client
account.

For each client account, you can configure one or more `[token <TOKEN_NAME>]`
sections, where `<TOKEN_NAME>` is a unique name of your choosing.
These sections describe how to create the token with the `oidc-token` tool.
For details, see
[Creating a new OIDC Client Account](#creating-a-new-oidc-client-account)
below.

Details for this configuration can be found below under

> [Configuring tokens for an OIDC client account](#configuring-tokens-for-an-oidc-client-account).

Reviewer: I think this entire section should be moved to `README.md`

  the client request with a passcode printed on the console.
  Follow the prompts (visit the web link, enter the request passcode,
  log in with your WLCG INDIGO IAM user account, and authorize the request).
- If this succeeds you will be prompted with a new
  `[account <ACCOUNT_SHORTNAME>]` section to add to your `config.ini`.
  Go ahead and add it, replacing the example section if it's still there.

Next you can configure one or more tokens for this client account.

Configuring tokens for an OIDC client account
---------------------------------------------

After you have created an OIDC client account and added it to
`/osg/token-renewer/config.ini`, you need to create a corresponding `token`
section in the config for each token that should be generated for this account
(possibly with different options).

Choose a `<TOKEN_NAME>` and add a new `[token <TOKEN_NAME>]` section
(replacing the example section if it's still there).

The `account` option in this section must match the `<ACCOUNT_SHORTNAME>`
for the corresponding `[account <ACCOUNT_SHORTNAME>]` section.

The `token_path` option specifies the output path where the generated token

> will be saved.

Author: So i think somewhere i mention that this is the default, and it is also spelled out like that in the example in the config.ini that gets installed with the package.  I guess the question is whether we want to let the users know that they can write tokens to other locations.  If not, i am happy with your version that just tells them what to put there.  Otherwise if they might want to write them elsewhere maybe we can present it as an option.

> will be saved.

Reviewer: I'd prefer to be prescriptive but I'm open to 

```
1.  Set the `token_path` to the output path where the generated token will be saved (recommended: `/etc/osg/tokens/<ACCOUNT_SHORTNAME>.token`) 
```

Optionally, you may also specify any of the following options, which will
be passed to the `oidc-token` command when generating the token:

| Option         | Description                                               |
|:---------------|:----------------------------------------------------------|
| `audience`     | list of audiences to pass via `--aud` option              |
| `scope`        | list of scopes to pass via `--scope` option               |
| `min_lifetime` | min token lifetime in seconds to pass via `--time` option |

!!! note
    For tokens used against an HTCondor-CE, set the `audience` option to

>     `<CE FQDN>:<CE PORT>`.

Reviewer: @edquist you fixed the spacing but I'd like the descriptions reworded slightly

Running the token renewal service
---------------------------------

The OSG token renewal service is set up as a "oneshot" systemd service,
which runs under the `osg-token-svc` user, sets up an `oidc-agent`,
adds the relevant OIDC client accounts as specified in the `config.ini`

> with `oidc-add`, and generates the tokens with `oidc-token`.

Reviewer: These are implementation details that should go into the README

> This service is set to run via a systemd timer nightly at midnight.

Reviewer: Add tables like here https://opensciencegrid.org/docs/data/stashcache/install-cache/#managing-stashcache-and-associated-services and here https://opensciencegrid.org/docs/data/stashcache/install-cache/#managing-stashcache-and-associated-services, listing `osg-token-renewer.service` and `osg-token-renewer.timer`

> If you would like to run the service manually at a different time (to generate

Reviewer: ```suggestion
If you would like to run the service manually at a different time (e.g., to generate
```
